Головнев Максим

maximka200 opened the pull request.

maximka200 committed:
- сделал основную структуру
- написал рабочий PrintToString(c оговорками)

maximka200 committed:
- написал тесты
- написал рабочий вариант(не сериализует коллекции)

maximka200 committed:
- дописал тестов
- реализовал принтинг коллекций(IEnumerable, IDictionary)

maximka200 commented:
@masssha1308

masssha1308 commented on ObjectPrinterTests/CollectionPrintingTests.cs:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    [Test]
+    public void IntArray_Snapshot()
```
Давай поправим названия тестов, так чтобы по названию было понятно что проверяет тест. Например, можно использовать такой `TestMethod_Condition_ExpectedResult` шаблон именования

masssha1308 commented on ObjectPrinterTests/CollectionPrintingTests.cs:
```diff
@@ -0,0 +1,151 @@
+{
+    private class Address
+    {
+        public string City { get; set; } = "";
```
Зачем выносить класс для одного поля?

masssha1308 commented on ObjectPrinterTests/CollectionPrintingTests.cs:
```diff
@@ -0,0 +1,151 @@
+namespace ObjectPrinterTests;
+
+[TestFixture]
+public class CollectionPrintingSnapshotTests
```
Стоит добавить тесты для граничных случаев (null-элементов, null-коллекций и т.д.)

masssha1308 commented on ObjectPrinterTests/Helper.cs:
```diff
@@ -0,0 +1,17 @@
+
+public static class Helper
+{
+    public static string Read(string fileName)
```
Какое преимущество перед готовыми инструментами? (Verify, ApprovalTests и т.д.). Если можно заиспользовать готовые решения, то лучше так и сделать

maximka200 replied:
> Какое преимущество перед готовыми инструментами? (Verify, ApprovalTests и т.д.). Если можно заиспользовать готовые решения, то лучше так и сделать

Писал до второй лекции о тестах, поэтому не знал о них. Переделаю :)

masssha1308 commented on ObjectPrinterTests/ObjectPrinterAcceptanceTests.cs:
```diff
@@ -0,0 +1,155 @@
+using FluentAssertions;
+using NUnit.Framework;
+
+namespace ObjectPrinting.Tests
```
namespace другой

masssha1308 commented on ObjectPrinterTests/ObjectPrinterAcceptanceTests.cs:
```diff
@@ -0,0 +1,155 @@
+        }
+    }
+    [TestFixture]
+    public class ObjectPrinterTests
```
Давай вынесем в отдельный файл, нет необходмости хранить эти классы в одном файле

masssha1308 commented on ObjectPrinterTests/ObjectPrinterAcceptanceTests.cs:
```diff
@@ -0,0 +1,155 @@
+    [TestFixture]
+    public class ObjectPrinterTests
+    {
+        private class A
```
По смыслу тестовые данные ObjectPrinterTests, ObjectPrinterAcceptanceTests и CollectionPrintingSnapshotTests похожи. Можно ли использовать общие тестовые данные?

maximka200 replied:
> По смыслу тестовые данные ObjectPrinterTests, ObjectPrinterAcceptanceTests и CollectionPrintingSnapshotTests похожи. Можно ли использовать общие тестовые данные?

По сути да, можно

masssha1308 commented on ObjectPrinterTests/ObjectPrinterAcceptanceTests.cs:
```diff
@@ -0,0 +1,155 @@
+        }
+        
+        [Test]
+        public void ShouldExcludeType_FromSerialization()
```
В названиях не хватает метода который проверяется

masssha1308 commented on ObjectPrinterTests/ObjectPrinterAcceptanceTests.cs:
```diff
@@ -0,0 +1,155 @@
+                .Excluding(p => p.Id)
+                .Excluding<Guid>();
+            
+            //7. Синтаксический сахар в виде метода расширения, сериализующего по-умолчанию  
```
Не понятно зачем эти комментарии + в этом тесте ничего не проверяется, текст просто выводится в консоль. Что должен проверять этот тест?

maximka200 replied:
Это тест из classwork, забыл удалить)

masssha1308 commented on ObjectPrinterTests/ObjectPrinterAcceptanceTests.cs:
```diff
@@ -0,0 +1,155 @@
+        }
+        
+        [Test]
+        public void Demo()
```
Название нужно поправить

masssha1308 commented on ObjectPrinterTests/ObjectPrinterAcceptanceTests.cs:
```diff
@@ -0,0 +1,155 @@
+            
+            var result = printer.PrintToString(obj);
+
+            result.Should().NotContain(nameof(A.Number));
```
Слабые проверки: проверяется только наличие/отсутствие полей, но при этом не проверяются остальные данные. Например, если excluding просто стирает все данные в файле, то по тестам мы это не узнаем и все тесты пройдут

masssha1308 commented on ObjectPrinterTests/ExpectedResults/Team_With_Members.txt:
```diff
@@ -0,0 +1,27 @@
+Team:
+	Title = My_Command
+	Members = List`1 [
```
В реализации сериализации не должно быть внутренних деталей языка (`List`1`, `Dictionary`2` и прочее)

masssha1308 commented on ObjectPrinting/PrintingConfig.cs:
```diff
@@ -1,41 +1,58 @@
-            }
-            return sb.ToString();
-        }
+    public string PrintToString(TOwner obj)
```
класс с настройками не должен печатать, это нарушение SRP. Давай либо логику, либо настройки вынесем в отдельный класс

masssha1308 commented on ObjectPrinting/Tests/Person.cs:
```diff
@@ -1,12 +0,0 @@
-﻿using System;
-
-namespace ObjectPrinting.Tests
```
Почему решил удалить эти тестовые данные и написать приватные классы?

maximka200 replied:
Чтобы можно было сделать Person с разными полями для разных типов тестов

> Почему решил удалить эти тестовые данные и написать приватные классы?

masssha1308 commented on ObjectPrinting/ObjectPrinter.cs:
```diff
@@ -1,10 +1,15 @@
+    public static string PrintToString<TOwner>(object obj)
+    {
+        var printer = new PrintingConfig<TOwner>();
+        return printer.PrintToString((TOwner)obj);
```
Почему в тестах используется не этот метод, а PrintingConfig.PrintToString? Для чего нужен этот метод?

maximka200 replied:
> Почему в тестах используется не этот метод, а PrintingConfig.PrintToString? Для чего нужен этот метод?

Это синтаксический сахар в виде метода расширения, сериализующего по-умолчанию, когда в классе  работали ребята просили реализовать

masssha1308 replied:
Сейчас он не используется поэтому его можно убрать

masssha1308 commented on fluent-api.sln.DotSettings:
```diff
@@ -1,6 +1,9 @@
 ﻿<wpf:ResourceDictionary xml:space="preserve" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml" xmlns:s="clr-namespace:System;assembly=mscorlib" xmlns:ss="urn:shemas-jetbrains-com:settings-storage-xaml" xmlns:wpf="http://schemas.microsoft.com/winfx/2006/xaml/presentation">
```
давай исключим этот файл из PR через gitignore, т.к. не несет полезной информации

masssha1308 commented on fluent-api.sln:
```diff
@@ -13,6 +13,8 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "FluentMapping.Tests", "Samp
 EndProject
```
давай исключим этот файл из PR через gitignore, т.к. генерируется автоматически и не несет полезной информации

masssha1308 commented on ObjectPrinting/PrintingConfig.cs:
```diff
@@ -1,41 +1,58 @@
-            };
-            if (finalTypes.Contains(obj.GetType()))
-                return obj + Environment.NewLine;
+    internal static string GetPropertyName<TProp>(Expression<Func<TOwner, TProp>> selector)
```
Нарушение SRP, давай этот метод вынесем в отдельный хэлпер

masssha1308 commented on ObjectPrinting/ObjectExtensions.cs:
```diff
@@ -0,0 +1,100 @@
+            var name = prop.Name;
+            var propType = prop.PropertyType;
+
+            if (config.ExcludedProperties.Contains(name) ||
```
Сильная связность, extension метод знает все детали реализации конфига. Такого быть не должно, т.к. малейшие изменения в конфиге повлекут за собой изменения в extension методе

masssha1308 commented on ObjectPrinting/ObjectExtensions.cs:
```diff
@@ -0,0 +1,100 @@
+
+public static class ObjectExtensions
+{
+    public static void Print<TOwner>(this object? obj, PrintingConfig<TOwner> config,
```
Получился god object который делает вообще всё: от проверки на null до сериализации коллекций + нарушение принципа OCP (чтобы добавить новый тип придется залазить в этот метод и менять код). Давай сделаем отдельные реализации сериализаторов для типов (nullable, словарь, default и т.д.) и реализуем паттерн стратегия. Должно получиться что-то вроде 
```
`public static void Print<TOwner>(this object? obj, PrintingConfig<TOwner> config, StringBuilder sb, int indent)
{
    foreach (var strategy in strategies)
    {
        if (strategy.CanPrint(obj, config))
        {
            strategy.Print(obj, config, sb, indent);
            return;
        }
    }
}`
```

В таком случае CollectionPrintingExtensions не понадобится

masssha1308 commented on ObjectPrinting/ObjectExtensions.cs:
```diff
@@ -0,0 +1,100 @@
+        PrintDefaultObject(obj, config, sb, indent);
+    }
+    
+    private static void PrintDefaultObject<TOwner>(object obj, PrintingConfig<TOwner> config,
```
Довольно запутанный метод получился, давай вынесем часть логики в отдельные методы для улучшения читабельности?

maximka200 committed:
Stop tracking fluent-api.sln and ignore it

maximka200 committed:
Stop tracking fluent-api.sln.DotSettings and ignore it

maximka200 committed:
- изменил gitignore

maximka200 committed:
- реализовал паттерн стратегия для ObjectPrinter.cs
- вынес всю лишнюю логику из PrintingConfig.cs в PrintingConfigExtensions.cs

maximka200 committed:
- переписал DefaultObjectStrategy, сделал более читаемым

maximka200 committed:
- поправил тест кейсы
- переписал реализацию парсинга коллекций(теперь выводится генерик-тип и нет лишних символов)

maximka200 committed:
- вернул fluent-api.sln

maximka200 committed:
добавил fluent-api.sln в .gitignore

maximka200 committed:
- переписал тесты с использованием Verify

maximka200 committed:
- еще тесты

masssha1308 commented on ObjectPrinting/Extensions/CollectionPrintingExtensions.cs:
```diff
@@ -0,0 +1,73 @@
+
+namespace ObjectPrinting.Extensions;
+
+public static class CollectionPrintingExtensions
```
Теперь можно избавиться от этого класса и перенести методы в классы-стратегии

masssha1308 commented on ObjectPrinterTests/ExpectedResults/CollectionPrintingSnapshotTests.PrintToString_ShouldHandleClassWithCollections_Correctly.verified.txt:
```diff
@@ -0,0 +1,20 @@
+		dev
+		qa
+	]
+	Addresses = Dictionary<String, Address> {
```
Хорошо бы переделать на языконезависимый JSON-подобный формат. Типы данных не должны быть явно прописаны, т.к. названия типов это детали C#. `List<String> [
		dev
		qa
	] -> ["dev", "qa"]`
